iree/samples/static_library/static_library_demo.c

```diff
@@ -0,0 +1,214 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+// A example of static library loading in IREE. See the README.md for more info.
+// Note: this demo requires artifacts from iree-translate before it will run.
+
+#include <stdio.h>
+
+#include "iree/hal/local/loaders/static_library_loader.h"
+#include "iree/hal/local/task_device.h"
+#include "iree/modules/hal/module.h"
+#include "iree/runtime/api.h"
+#include "iree/samples/static_library/simple_mul_c.h"
+#include "iree/task/api.h"
+#include "iree/vm/bytecode_module.h"
+
+// Compiled static library module here to avoid IO:
+#include "iree/samples/static_library/simple_mul.h"
+
+// A function to create the HAL device from the different backend targets.
+// The HAL device is returned based on the implementation, and it must be
+// released by the caller.
+iree_status_t create_device_with_static_loader(iree_hal_device_t** device) {
+  iree_status_t status = iree_ok_status();
+  // Set paramters for the device created in the next step.
+  iree_hal_task_device_params_t params;
+  iree_hal_task_device_params_initialize(&params);
+
+  // Load the statically embedded library
+  const iree_hal_executable_library_header_t** static_library =
+      simple_mul_dispatch_0_library_query(
+          IREE_HAL_EXECUTABLE_LIBRARY_LATEST_VERSION, /*reserved=*/NULL);
+  const iree_hal_executable_library_header_t** libraries[1] = {static_library};
+
+  iree_hal_executable_loader_t* library_loader = NULL;
+  if (iree_status_is_ok(status)) {
+    status = iree_hal_static_library_loader_create(
+        IREE_ARRAYSIZE(libraries), libraries,
+        iree_hal_executable_import_provider_null(), iree_allocator_system(),
+        &library_loader);
+  }
+
+  iree_task_executor_t* executor = NULL;
+  if (iree_status_is_ok(status)) {
+    iree_task_executor_create_from_flags(iree_allocator_system(), &executor);
+  }
+
+  // Create the device and release the executor and loader afterwards.
+  if (iree_status_is_ok(status)) {
+    iree_hal_task_device_create(iree_make_cstring_view("dylib"), &params,
+                                executor, 1, &library_loader,
+                                iree_allocator_system(), device);
+  }
+  iree_task_executor_release(executor);
+  iree_hal_executable_loader_release(library_loader);
+
+  return status;
+}
+
+iree_status_t Run() {
+  iree_status_t status = iree_ok_status();
```

Comment:
since this is a sample not intended to show proper use of the APIs but instead a specific build configuration it'd be fine to avoid the status handling and use `IREE_CHECK_OK` everywhere - this would keep the code more readable, and as long as it's called out at the top that this is for reference of how to use a static library (vs how this would integrate into a real application) it's fine. up to you.

Reply:
SG. Since I already changed it based on Scott's earlier comment, I'll think I'll leave it for now. I've addressed other comments to ensure usage is correct.